src/gordon_gcp/clients/http.py

```diff
@@ -95,15 +92,96 @@ def __init__(self, auth_client=None, session=None):
             headers (dict): (optional) HTTP headers to send with the
-                request. Headers pass through to the request will
+                request. Headers passed through to the request will
                 include :attr:`DEFAULT_REQUEST_HEADERS`.
+            retry_type (str): (optional) The type of retries to use: "none",
+                "simple", (retry up to tries/timeout), "no4xx" (don't retry
+                on 4xx, otherwise up to tries/timeout), or "custom".  If
+                "custom", uses the retry predicate specified in the other
+                parameters.  See also extra_401_retry.
+            tries (int): (optional) Total maximum number of times to try the
+                request, including the first (i.e. 1 == no retries).  The number
+                of actual tries is subject to retry_type and timeout as well as
+                this value.
+            timeout (int): (optional) Total maximum seconds to keep trying the
+                request (None == unlimited).  The number of actual tries is
+                subject to retry_type and tries as well as this value.
+            retry_predicate (callable): (optional) Function to call to decide
+                if we should keep trying the request.  Ignored unless
+                retry_type is "custom".  Must return True for "continue retrying
+                (subject to tries and timeout)" or False for "stop retrying".
+                Last argument must be the response from _request().  If the
+                function takes any other arguments, their values must be
+                supplied in predicate_args and/or predicate_kwargs.
+            predicate_args (list): (optional) See retry_predicate.
+            predicate_kwargs (dict): (optional) See retry_predicate.
+            extra_401_retry (bool): (optional) If True, run one more request
+                attempt on a 401 (unauthorized), regardless of the value of
+                retry_type.  This makes sure we've tried generating a new token.
         Returns:
             (str) HTTP response body.
         Raises:
-            :exc:`.GCPHTTPError`: if any exception occurred.
+            :exc:`.GCPHTTPError`: If a response code >=400 was received and
+                there are no more tries (or no retrying is specified).
         """
-        refresh_attempt = kwargs.pop('cred_refresh_attempt', 0)
+        retry_predicates = {
+            "none": lambda resp: False,  # never retry
+            # retry until success, subject to tries/timeout
+            "simple": lambda resp: resp.status >= 400,
+            # don't retry on 4xx
+            "no4xx": lambda resp: resp.status >= 500,
+            "custom": retry_predicate,
+        }
+        actual_predicate = retry_predicates[retry_type]
+        if predicate_args is None:
+            predicate_args = []
+        if predicate_kwargs is None:
+            predicate_kwargs = {}
+        predicate_partial = partial(
```

Comment:
(again, if we end up going down this route) I think the user of this method should construct their own partial and supply this, since it's not relevant to the non-custom predicates.

Reply:
I can do that.  I just wasn't sure if it's putting too much work on the caller, vs just centralizing the partial() call.